src/mirror/mod.rs

```diff
@@ -0,0 +1,143 @@
+//! This is a temporary workaround to let us combine multiple topics into a single stream so
+//! consumers work properly. This will be removed once NATS 2.10 is out and we have upgraded to
+//! using it in wasmcloud projects
+
+use std::{collections::HashMap, sync::Arc};
+
+use async_nats::{
+    jetstream::{
+        consumer::pull::{Config as PullConfig, Stream as MessageStream},
+        stream::Stream as JsStream,
+        Context,
+    },
+    Error as NatsError, HeaderMap,
+};
+use bytes::Bytes;
+use futures::StreamExt;
+use tokio::{sync::RwLock, task::JoinHandle};
+use tracing::{error, instrument, trace, warn};
+
+type WorkHandles = Arc<RwLock<HashMap<String, JoinHandle<anyhow::Result<()>>>>>;
+
+/// A simple NATS consumer that takes each incoming message and maps it to `{prefix}.{lattice-id}`
+/// (e.g. `wadm.evt.default`)
+pub struct Mirror {
+    stream: JsStream,
+    prefix: String,
+    handles: WorkHandles,
+}
+
+impl Mirror {
+    /// Returns a new Mirror for the given stream
+    pub fn new(stream: JsStream, prefix: &str) -> Mirror {
+        Mirror {
+            stream,
+            prefix: prefix.to_owned(),
+            handles: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    #[instrument(level = "debug", skip(self))]
+    pub async fn monitor_lattice(&self, subject: &str, lattice_id: &str) -> Result<(), NatsError> {
+        if let Some(handle) = self.handles.read().await.get(lattice_id) {
+            if !handle.is_finished() {
+                return Ok(());
+            }
+            warn!("Handle was marked as completed. Starting monitor again");
+        }
+        let consumer_name = format!("wadm_mirror_{lattice_id}");
+        trace!("Creating mirror consumer for lattice");
+        let consumer = self
+            .stream
+            .get_or_create_consumer(
+                &consumer_name,
+                PullConfig {
+                    durable_name: Some(consumer_name.clone()),
+                    name: Some(consumer_name.clone()),
+                    description: Some(format!("Durable wadm mirror consumer for {lattice_id}")),
+                    ack_policy: async_nats::jetstream::consumer::AckPolicy::Explicit,
+                    // These should be real quick to ack
+                    ack_wait: std::time::Duration::from_secs(1),
```

Comment:
Since this is a mirror, it's always a local NATS message right? I'd only ask to make this configurable if this message will traverse over a real network boundary

Reply:
It will likely cross a network boundary. I'll bump this to 2s, but because this is temporary I don't really see a need to make it configurable